ocaml/xapi/xapi_pool_helpers.ml

```diff
@@ -269,3 +269,221 @@ let apply_guest_agent_config ~__context =
         (Printexc.to_string e)
   in
   call_fn_on_slaves_then_master ~__context f
+
+module PeriodicUpdateSync = struct
+  let periodic_update_sync_task_name = "Periodic update synchronization"
+
+  exception UpdateSync_RetryNumExceeded of int
+
+  let seconds_random_within_a_day () =
+    if Xapi_fist.disable_periodic_update_sync_sec_randomness () then
+      0.
+    else
+      let secs_of_a_day = 24. *. 60. *. 60. in
+      let secs_random = Random.float secs_of_a_day in
+      secs_random
+
+  let frequency_to_str ~frequency =
+    match frequency with
+    | `daily ->
+        "daily"
+    | `monthly ->
+        "monthly"
+    | `weekly ->
+        "weekly"
+
+  let weekday_to_int = function
+    | `Sun ->
+        0
+    | `Mon ->
+        1
+    | `Tue ->
+        2
+    | `Wed ->
+        3
+    | `Thu ->
+        4
+    | `Fri ->
+        5
+    | `Sat ->
+        6
+
+  let utc_start_of_next_scheduled_day ~utc_now ~tz_offset_s ~frequency
+      ~day_configed_int =
+    let y, m, d = fst (Ptime.to_date_time ~tz_offset_s utc_now) in
+    let utc_start_of_today =
+      Ptime.of_date_time ((y, m, d), ((0, 0, 0), tz_offset_s)) |> Option.get
```

Comment:
`Ptime.of_date_time ~tz_offset_s (y, m, d) |> Option.get`

Reply:
The TZ parameter is not supported in the version of PTime we are using.(we are not using the latest version)